File: CourseProjects/CourseProjects/VisualStudioProjects/Agile/MockingWithMoq/MockingChallengeFinish/TravelAllowanceCalculatorTest.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace MockingChallengeFinish {

    [TestClass]
    public class TravelAllowanceCalculatorTest {
        [TestInitialize]
        public void Start() {
            concessionsEngine = new Mock<IFlightConcessionsEngine>();
            pathDatabase = new Mock<IFlightPathDatabase>();
            calculator = new TravelAllowanceCalculator(concessionsEngine.Object, pathDatabase.Object);
        }
        [TestMethod]
        public void LocalFlightFirstClass() {
            Flight f = new Flight("Belfast", "Dublin", Seating.FIRST, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Belfast", "Dublin")).Returns(90);

            Assert.AreEqual(9.0, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 9.0),Times.Once());
        }
        [TestMethod]
        public void LocalFlightBusinessClass() {
            Flight f = new Flight("Belfast", "Dublin", Seating.BUSINESS, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Belfast", "Dublin")).Returns(90);

            Assert.AreEqual(27.0, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 27.0), Times.Once());
        }
        [TestMethod]
        public void LocalFlightEconomyClass() {
            Flight f = new Flight("Belfast", "Dublin", Seating.ECONOMY, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Belfast", "Dublin")).Returns(90);

            Assert.AreEqual(45.0, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 45.0), Times.Once());
        }
        [TestMethod]
        public void InternationalFlightFirstClass() {
            Flight f = new Flight("Heathrow", "Dublin", Seating.FIRST, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Heathrow", "Dublin")).Returns(350);

            Assert.AreEqual(70.0, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 70.0), Times.Once());
        }
        [TestMethod]
        public void InternationalFlightBusinessClass() {
            Flight f = new Flight("Heathrow", "Dublin", Seating.BUSINESS, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Heathrow", "Dublin")).Returns(350);

            Assert.AreEqual(140.0, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 140.0), Times.Once());
        }
        [TestMethod]
        public void InternationalFlightEconomyClass() {
            Flight f = new Flight("Heathrow", "Dublin", Seating.ECONOMY, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Heathrow", "Dublin")).Returns(350);

            Assert.AreEqual(210, calculator.CalculateAllowance(f), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(1, 210.0), Times.Once());
        }
        [TestMethod]
	    public void MultipleLocalFlights() {
            Flight f1 = new Flight("Belfast", "Dublin", Seating.FIRST, DateTime.Now);
            Flight f2 = new Flight("Belfast", "Dublin", Seating.FIRST, DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Belfast", "Dublin")).Returns(90);

            Assert.AreEqual(18, calculator.CalculateAllowance(f1, f2), 0.001);
            concessionsEngine.Verify(m => m.AdditionalBenefitApplies(2, 18.0), Times.Once());
	    }
        [TestMethod]
	    //Multiple international flights should qualify for extra allowance
	    public void MultipleInternationalFlights() {
            Flight f1 = new Flight("Heathrow", "Dublin", Seating.FIRST, DateTime.Now);
            Flight f2 = new Flight("Heathrow", "Dublin", Seating.FIRST, DateTime.Now);
		    Flight f3 = new Flight("Heathrow","Dublin",Seating.FIRST,DateTime.Now);

            pathDatabase.Setup(m => m.FindFlightPath("Heathrow", "Dublin")).Returns(350);
            concessionsEngine.Setup(m => m.AdditionalBenefitApplies(3, 210.0)).Returns(true);
            concessionsEngine.Setup(m => m.CalculateBenefit(3, 210.0)).Returns(25.0);

		    Assert.AreEqual(235.0,calculator.CalculateAllowance(f1,f2,f3),0.001);
            pathDatabase.Verify(m => m.FindFlightPath(It.IsAny<string>(), It.IsAny<string>()),Times.Exactly(3));
	    }
        private TravelAllowanceCalculator calculator;
        private Mock<IFlightConcessionsEngine> concessionsEngine;
        private Mock<IFlightPathDatabase> pathDatabase;
    }
}

File: CourseProjects/CourseProjects/JavaEclipseProjects/UnitTestingExercises/src/exercise/tdd/mocking/finish/mockito/TravelAllowanceCalculatorTest.java
package exercise.tdd.mocking.finish.mockito;

import static exercise.tdd.mocking.finish.mockito.Seating.*;
import static org.junit.Assert.*;
import java.util.Date;

import org.junit.*;
import static org.mockito.Mockito.*;

public class TravelAllowanceCalculatorTest {
	@Before
	public void start() {
		concessionsEngine = mock(FlightConcessionsEngine.class);
		pathDatabase = mock(FlightPathDatabase.class);
		calculator = new TravelAllowanceCalculator(concessionsEngine,pathDatabase);
	}
	@Test
    public void localFlightFirstClass() {
        Flight f = new Flight("Belfast","Dublin",FIRST,new Date());
        when(pathDatabase.findFlightPath("Belfast","Dublin")).thenReturn(90L);
        when(concessionsEngine.additionalBenefitApplies(1, 9.0)).thenReturn(false);

        assertEquals(9.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
    public void localFlightBusinessClass() {
        Flight f = new Flight("Belfast","Dublin",BUSINESS,new Date());
        when(pathDatabase.findFlightPath("Belfast","Dublin")).thenReturn(90L);
        when(concessionsEngine.additionalBenefitApplies(1, 27.0)).thenReturn(false);

        assertEquals(27.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
    public void localFlightEconomyClass() {
        Flight f = new Flight("Belfast","Dublin",ECONOMY,new Date());
        when(pathDatabase.findFlightPath("Belfast","Dublin")).thenReturn(90L);
        when(concessionsEngine.additionalBenefitApplies(1, 45.0)).thenReturn(false);

        assertEquals(45.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
    public void internationalFlightFirstClass() {
        Flight f = new Flight("Heathrow","Dublin",FIRST,new Date());
        when(pathDatabase.findFlightPath("Heathrow","Dublin")).thenReturn(350L);
        when(concessionsEngine.additionalBenefitApplies(1, 70.0)).thenReturn(false);

        assertEquals(70.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
    public void internationalFlightBusinessClass() {
        Flight f = new Flight("Heathrow","Dublin",BUSINESS,new Date());
        when(pathDatabase.findFlightPath("Heathrow","Dublin")).thenReturn(350L);
        when(concessionsEngine.additionalBenefitApplies(1, 140.0)).thenReturn(false);

        assertEquals(140.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
    public void internationalFlightEconomyClass() {
        Flight f = new Flight("Heathrow","Dublin",ECONOMY,new Date());
        when(pathDatabase.findFlightPath("Heathrow","Dublin")).thenReturn(350L);
        when(concessionsEngine.additionalBenefitApplies(1, 210.0)).thenReturn(false);
 
        assertEquals(210.0,calculator.calculateAllowance(f),0.001);
    }
	@Test
	public void multipleLocalFlights() {
		Flight f1 = new Flight("Belfast","Dublin",FIRST,new Date());
		Flight f2 = new Flight("Belfast","Dublin",FIRST,new Date());
		
		when(pathDatabase.findFlightPath("Belfast","Dublin")).thenReturn(90L);
        when(concessionsEngine.additionalBenefitApplies(2, 18.0)).thenReturn(false);
 
		assertEquals(18.0,calculator.calculateAllowance(f1,f2),0.001);
		verify(pathDatabase,times(2)).findFlightPath(anyString(),anyString());
	}
	@Test
	//multiple international flights should qualify for extra allowance
	public void multipleInternationalFlights() {
		Flight f1 = new Flight("Heathrow","Dublin",FIRST,new Date());
		Flight f2 = new Flight("Heathrow","Dublin",FIRST,new Date());
		Flight f3 = new Flight("Heathrow","Dublin",FIRST,new Date());
		
		when(pathDatabase.findFlightPath("Heathrow","Dublin")).thenReturn(350L);
		when(concessionsEngine.additionalBenefitApplies(3, 210.0)).thenReturn(true);
		when(concessionsEngine.calculateBenefit(3, 210.0)).thenReturn(25.0);

		assertEquals(235.0,calculator.calculateAllowance(f1,f2,f3),0.001);
		verify(pathDatabase,times(3)).findFlightPath(anyString(),anyString());
	}
	private TravelAllowanceCalculator calculator;
	private FlightConcessionsEngine concessionsEngine;
	private FlightPathDatabase pathDatabase;
}
